### ISSUE TITLE
Add the ability to trigger action from pro-code

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-toolpad#351]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

I extract an order id from a license key in https://master--toolpad.mui.com/_toolpad/app/cl1c2j0l512939zo6a43575en/editor/pages/cl2kfa28r00013f69amqv1prv. However, at this point, I would now like to call the WooCommerce API with this order id and display more information about the order. I assume this requires being able to call a query.

### Examples 🌈

Retool has the following API:

```jsx
<style>
  body {
    margin: 0;
  }
</style>
<script src="https://cdn.tryretool.com/js/react.production.min.js" crossorigin></script>
<script src="https://cdn.tryretool.com/js/react-dom.production.min.js" crossorigin></script>
<script src="https://unpkg.com/@material-ui/core@3.9.3/umd/material-ui.production.min.js"></script>


<div id="react"></div>

<script type="text/babel">
  const { Button, Card, CardContent } = window["material-ui"];

  const MyCustomComponent = ({ triggerQuery, model, modelUpdate }) => (
    <Card>
      <CardContent>
        <div>{model.displayText}</div>
        <Button
          color="primary"
          variant="outlined"
          onClick={() => triggerQuery(model.queryToTrigger)}
        >
          Trigger {model.queryToTrigger}
        </Button>
        <Button
          color="primary"
          variant="outlined"
          onClick={() =>
            modelUpdate({ displayText: "much custom, such flexibility. wow." })
          }
        >
          Update Model
        </Button>
      </CardContent>
    </Card>
  );

  const ConnectedComponent = Retool.connectReactComponent(MyCustomComponent);
  ReactDOM.render(<ConnectedComponent />, document.getElementById("react"));
</script>
```

### Motivation 🔦

Accepting props is great, but custom components also need to interact with the rest of the app.

**Comment by Janpot** [comment]
We'll most likely solve that with some hooks, e.g.

```ts
import { useQuery } from '@mui/toolpad'
```

**Comment by oliviertassinari** [comment]
@Janpot This works for me.

Staying on this same topic of "feedback loop". I also wonder if we shouldn't make it possible for a component to have a global state so that it can be used as the other components. For instance, this component exposes a density.value

<img width="149" alt="Screenshot 2022-05-01 at 18 22 52" src="https://user-images.githubusercontent.com/3165635/166155016-8d9d71bb-65b3-4b60-b36f-fdda356ed9f9.png">

**Comment by Janpot** [comment]
The main problem to solve with that will be order of evaluation. When does which binding expression evaluate, how do you deal with cyclic structures, and how to detect infinite loops?

**Comment by newguy-123** [comment]
Retool exposes 3 variables for a Custom component to interact with other components. Details here:
https://docs.retool.com/docs/custom-react-components
Will we follow a similar approach?

**Comment by Janpot** [comment]
We will most likely expose hooks to interact with the toolpad environment. For example:

```tsx
import { usePageLink } from '@mui/toolpad';
import { Link } from '@mui/material'

export default function MyButton () {
  const href = usePageLink('other-page-id', { someVar: 'foo' })
  return <Link href={href}>To other the page!</Link>
}
```

Or:

```tsx
import { useQuery } from '@mui/toolpad';
import { Box , CircularProgress } from '@mui/material'

export default function MyData () {
  const { loading, data } = useQuery('some-query-id', { someVar: 'foo' })
  return loading ? <CircularProgress /> : <Box component='pre'>{JSON.stringify(data, null, 2)}</Box>
}
```

**Comment by Janpot** [comment]
@teowave Have you already checked out [Toolpad data providers](https://mui.com/toolpad/studio/concepts/data-providers/#row-creation). They have the ability add all CRUD operations to a datagrid. If this doesn't work for you, could you elaborate a bit more in your exact use-case?

**Comment by teowave** [comment]
thank you for the reply, Jan

yes, it works well with grid, but how can i for example create record by using it in my custom component. Is there an example somewhere that I can follow? 

Many thanks

**Comment by teowave** [comment]
<img width="811" alt="image" src="https://github.com/mui/mui-toolpad/assets/107956897/cb553708-4762-490e-9a04-8b1722294af7">

**Comment by Janpot** [comment]
@teowave Yes, that makes sense. I'm opening https://github.com/mui/mui-toolpad/issues/3357. We're also planning to open up this API for https://github.com/mui/mui-toolpad/discussions/3311

**Comment by bharatkashyap** [comment]
Might also be beneficial to check the [RFC on `@toolpad/core`](https://github.com/mui/mui-toolpad/discussions/3311) given your use case
